Hamiltonian Monte Carlo

khalibartan opened the pull request:
This PR deals with implementing HMC with dual averaging.
The implementation is still open for discussion. If you find anything ambiguous please comment on the line. This PR is open for discussion.

khalibartan committed:
Hamiltonian Monte Carlo

Implemented euler method and leapfrog method

ankurankan commented on pgmpy/inference/continuous_sampling.py:
```diff
@@ -0,0 +1,182 @@
+
+        self.log_grad_pdf = log_grad_pdf
+
+    def leapfrog(self, theta, momentum, epsilon):
```
@khalibartan The various methods should be separate classes (and you should be passing the classes as argument for `split_time`) with a common base class, so that anyone can write a new method and use it anywhere in pgmpy.

ankurankan replied:
@khalibartan These methods should be separate classes (and should be passing classes to `split_time` argument) with a common base class. Will help in users to quickly write their own methods and plug it in pgmpy.

khalibartan replied:
@ankurankan Will do that.

yashu-seth commented:
@khalibartan Will these algorithms work on multivariate gaussaian distribution only? Or they work on any class of multivariate distribution?

yashu-seth commented:
@khalibartan I was going through your timeline. I saw that in week 10 and 11 you are planning to rewrite the existing code to use models instead of explicitly specified parameters. Can you tell me about what you are planning? Since I will be working on the multivariate distribution class, I think we can decide some things beforehand. It can save a lot of effort.

khalibartan commented:
@yashu-seth It can work even with other type of continuous distributions like HMMs and mixtures. But since I don't know how they will be implemented, I started by using simplest multivariate gaussian and later allocated the time to generalize it. The sole issue while generalization is how to find gradient of posterior distribution and  gradient log of posterior distribution . Can you provide link to your gsoc proposal draft, than I'll be able to give a better comment about things for multivariate distribution. I did planned something for it but cannot remember exactly.

yashu-seth commented:
@khalibartan My proposal link - https://github.com/yashu-seth/gsoc

So the HMC and no U-turn can work only with multivariate gaussian?

khalibartan commented:
@yashu-seth No they can work with all kind of continuous models as stated here

> It can work even with other type of continuous distributions like HMMs and mixtures.

yashu-seth commented:
@khalibartan No, I meant for now you are implementing it for gaussian distribution only? We are still to figure its use with other models?

khalibartan commented:
@yashu-seth Yes, for now I'm implementing it accordingly to gaussian models, for I don't know how other models will be implemented. As I mentioned in my earlier comment the issue while generalization is how to find gradient of the probability distribution. Otherwise the same implementation can work with other models too.

yashu-seth commented:
@khalibartan Okay. Please go through the `JointGaussianDistribution` class in my proposal, it will give you an idea of the implementation I have planned.

khalibartan commented:
@yashu-seth Thanks! Yes I went through it, I feel there shouldn't be any difficulty as we can get `mean vector` and `covariance matrix` from the model itself.

khalibartan committed:
Created base classes for user defined methods

yashu-seth commented:
@khalibartan Cool.

yashu-seth commented:
@khalibartan Hey, I was thinking that we can have all the continuous part of the library in a separate directory. I have put ContinuousFactor.py in pgmpy/continuous/factors but made sure that it can be imported by `from pgmpy.factors import ContinuousFactor` . I think the sampling classes can be put  inside pgmpy/continuous/inference as well. What do you think about this?

khalibartan commented:
@yashu-seth Agreed! Even I was thinking of that (thought we will discuss that in this week meeting). We can create `pgmpy.discrete` and `pgmpy.continuous` kind of hierarchy and then categorizing remaining code accordingly  . It seems better that way.

khalibartan commented on pgmpy/inference/base_continuous.py:
```diff
@@ -0,0 +1,212 @@
+﻿#!/usr/bin/env python3
```
@ankurankan I have tried to create base classes. Please check if this what you meant. And for example should I demonstrate how to create a class using these base class or something else like in base Inference.

khalibartan committed:
Added the _find_resonable_epsilon method

khalibartan commented:
@ankurankan Please look at the changes. I have created a `AbstractGaussian` class because it was getting quite cumbersome to handle them as mean_vec and cov_matrix. Will remove that class later. I have created a `BaseHMC` class, will add three more methods to find `kinetic energy`, `potential energy` and `hamiltonian` . Also check for the base class implementation of `DiscretizeTime` and `GradientLogPDF` .

khalibartan commented:
@ankurankan Please look at the changes. I have created a `AbstractGaussian` class because it was getting quite cumbersome to handle them as mean_vec and cov_matrix. Will remove that class later. I have created a `BaseHMC` class, will add three more methods to find `kinetic energy`, `potential energy` and `hamiltonian` . Also check for the base class implementation of `DiscretizeTime` and `GradientLogPDF` .

khalibartan committed:
Created a AbstractGaussian class to take model parameter

khalibartan commented:
@ankurankan  @yashu-seth Please look at the changes. I have created a `AbstractGaussian` class because it was getting quite cumbersome to handle them as mean_vec and cov_matrix. Will remove that class later . Also check for the base class implementation of `DiscretizeTime` and `GradientLogPDF`  if it is what you meant by creating a base class.

khalibartan commented:
@ankurankan  @yashu-seth Please look at the changes. I have created a `AbstractGaussian` class just to have a model parameter(Was just irritated to see `mean_vec` and `cov_matrix`). Will remove that class later . Also check for the base class implementation of `DiscretizeTime` and `GradientLogPDF`  if it is what you meant by creating a base class. Any other suggestions also please comment them at necessary line.

yashu-seth commented on pgmpy/inference/base_continuous.py:
```diff
@@ -0,0 +1,343 @@
+    """
+
+    def __init__(self, grad_log_pdf, model, theta, momentum, epsilon):
+        # TODO: Take model instead of precision matrix
```
@khalibartan I think you have already taken model as input parameter. This comment can be removed.

yashu-seth commented on pgmpy/inference/base_continuous.py:
```diff
@@ -0,0 +1,343 @@
+
+    def __init__(self, model, grad_log_pdf,
+                 split_time=LeapFrog, delta=0.65):
+        # TODO: Use model instead of mean_vec and cov_matrix
```
@khalibartan Same here.

yashu-seth commented:
@khalibartan I think you can rename the `AbstractGaussian` class to `JointGaussainDistribution`. It will save making changes later to wherever you have mentioned it in the doc-strings.  Rest of the things are exactly as I have planned. So no other major changes would be necessary.

According to my timeline I was suppose to work on `LinearGaussainCPD` in the next week. But I will start working on `JointGaussainDistribution` instead and try to complete it as soon as possible.

khalibartan committed:
Fixed _find_resonable_epsilon

khalibartan commented:
@ankurankan How should I return the samples, as we no longer intend to support, pandas data frames.

khalibartan committed:
Added the sampling method

khalibartan committed:
Fixes in sample method of HMCda

khalibartan commented:
@ankurankan I have implemented the sample, please look at it. I have also tested it and it is working fine. If things look good to you I'll start writing tests for it.

khalibartan commented:
@ankurankan I have implemented the `sample method`, please look at it. I have also tested it and it is working fine. If things look good to you I'll start writing tests for it.

khalibartan committed:
Added the test for HMC sampling

khalibartan commented:
@ankurankan I have added the test too. Look at the things now. If everything is fine with you I'll add more tests and complete the remaining docstrings and final touch ups.

ankurankan commented on pgmpy/inference/base_continuous.py:
```diff
@@ -0,0 +1,308 @@
+        return grad, log_pdf
+
+
+class DiscretizeTime(object):
```
@khalibartan I think the name `DiscretizeTime` is too misleading.

khalibartan replied:
@ankurankan What should we name it then, Earlier I was using `SplitTime` .

ankurankan commented:
@khalibartan Always name base classes starting with `Base`

ankurankan commented on pgmpy/inference/base_continuous.py:
```diff
@@ -0,0 +1,308 @@
+
+    def __init__(self, grad_log_pdf, model, theta, momentum, epsilon):
+        # TODO: Take model instead of precision matrix
+        if isinstance(theta, (np.matrix, np.ndarray, list, tuple, set, frozenset)):
```
@khalibartan Google for a better way of doing this.

khalibartan replied:
@ankurankan I think this is the cleanest way for doing this. Though I'll look for others also.

ankurankan commented on pgmpy/inference/continuous_sampling.py:
```diff
@@ -0,0 +1,302 @@
+
+        return epsilon_app
+
+    def sample(self, theta0, num_adapt, num_samples, epsilon=None):
```
@khalibartan Since you have used a for loop in this function, I think you should be able to write `generate_sample` using this method.

ankurankan replied:
@khalibartan Hadn't we discussed about `sample` method accepting an argument for which method to use for computing steps (LeapFrog or Euler).

khalibartan replied:
@ankurankan I have taken that argument [here](https://github.com/pgmpy/pgmpy/pull/691/files#diff-2bd8b8b832815e6d408ffeb0a4b70837R27) in the `__init__` method. Should I take it in sample method ?

khalibartan replied:
@ankurankan I have copy pasted the code, in generate_sample method and made changes in one or two lines. Now I think I can use generate_sample code to write sample method code. Can I do that ?

ankurankan commented on pgmpy/inference/continuous_sampling.py:
```diff
@@ -0,0 +1,302 @@
+
+        References
+        ----------
+        Matthew D. Hoffman, Andrew Gelman, The No-U-Turn Sampler: Adaptively
```
@khalibartan Is this code for NUTS?

khalibartan replied:
@ankurankan No, but it uses algorithm for Hamiltonian Monte Carlo, given in that paper. Below I have mentioned the algorithm I have used from the paper.

khalibartan commented:
@ankurankan How should I take the parameter theta as input.

khalibartan committed:
Added implementation of Simple HMC, replaced np.matrix with np.array

khalibartan commented on pgmpy/inference/continuous_sampling.py:
```diff
@@ -0,0 +1,579 @@
+        if num_adapt <= 1:  # return sample generated using Simple HMC algorithm
+            for sample in HamiltonianMC.generate_sample(self, theta0, num_samples, epsilon, Lambda):
+                yield sample
+            return
```
@ankurankan Can you suggest a better solution for this ? Python 2.X doesn't allow to return generator objects.

khalibartan commented on pgmpy/inference/base_continuous.py:
```diff
@@ -0,0 +1,268 @@
+        return grad, log_pdf
+
+
+class BaseDiscretizeTime(object):
```
@ankurankan Better name for this class, though in research papers and book euler and leapfrog methods have been called methods to discretize time

khalibartan commented on pgmpy/inference/continuous_sampling.py:
```diff
@@ -0,0 +1,579 @@
+        array([[ 0.64321553,  0.63513749],
+               [ 0.63513749,  0.98544953]])
+        """
+        # TODO: Proper parameterization
```
@ankurankan This TODO needs to be resloved. Can you suggest how should I take parameter theta.

khalibartan commented:
@ankurankan Please look at the inline comments. Earlier response will help me finish up HMC by end of this week.

khalibartan committed:
Replaced acceptance prob by function

khalibartan commented:
@ankurankan Can we take parameter theta as list of n dimensional arrays, flatten them into one flat array, create samples using it and then reshaping them upon return

ankurankan commented:
@khalibartan Sorry for the delay. Is this an absolute blocker right now ?

khalibartan commented:
@ankurankan Apart from this issue implementation is complete, just a bit of tests for exception cases, other than that every thing seems right to me.

ankurankan commented:
@khalibartan Okay. Then can you start working on the tests and NUTS? I will need a couple of days on this.

khalibartan commented:
@ankurankan I was thinking of writing a cache implementation, going through algorithm I found that we evaluate same gradient repeatedly sometimes, and gradient computation can be difficult for complex models, so I think a cache implementation might be helpful in here, also I was thinking of adding a verbose argument. What are your suggestions ?

ankurankan commented:
@khalibartan Yeah you can do that.

ankurankan commented on pgmpy/inference/base_continuous.py:
```diff
@@ -0,0 +1,268 @@
+
+    Parameters
+    ----------
+    theta : A 1d array type object or a row matrix of shape 1 X d
```
@khalibartan Check for the way we have written docstrings everywhere.

ankurankan commented on pgmpy/inference/base_continuous.py:
```diff
@@ -0,0 +1,268 @@
+    >>> from pgmpy.inference import JointGuassianDistribution as JGD
+    >>> from pgmpy.inference import BaseGradLogPDF
+    >>> import numpy as np
+    >>> class GradLogGaussian(BaseGradLogPDF):
```
@khalibartan What is the point of showing this in the doctests?

khalibartan replied:
@ankurankan I believe docstring are intended for user to know how to use particular class or function. I have added more things to it, now it also shows how to use it for finding proposal values.

ankurankan commented on pgmpy/inference/base_continuous.py:
```diff
@@ -0,0 +1,268 @@
+
+    def _get_gradient_log_pdf(self):
+        """
+        Method that finds gradient and its log at theta
```
Add proper docstring

ankurankan commented:
@khalibartan If I am understanding correctly theta is the position vector (And you shouldn't name something theta rather name should tell us what it is). Right ? If it's the position vector I don't think you should be fine with a 1D array.

ankurankan commented on pgmpy/inference/base_continuous.py:
```diff
@@ -0,0 +1,268 @@
+        self.precision_matrix = np.linalg.inv(cov_matrix)
+
+
+class BaseGradLogPDF(object):
```
@khalibartan Do you think gradient computation should be a method in `ContinuousFactor` or classes inheriting it ?

khalibartan replied:
@ankurankan Gradient computation should be in models i.e class inheriting it. Is there an issue with name. I'm not good at naming things.

khalibartan commented:
@ankurankan I used the name theta, because it was used in algorithm, I'll change it to position then.

> If it's the position vector I don't think you should be fine with a 1D array.

Yes indeed! I think we will need multi-dimensional array, but the product `(np.dot)` doesn't function like we intend to have it here in case of higher dimensions, so I propose that we flatten these values to a uni-dimensional array, use this uni-dimensional array and apply algorithm on it, and later reconstruct them back in proper shape.

If things can be parametrized only in 2 dimensions, than there won't be much of issue. But I'm really sure of this.

Also there was another thing I wanted to talk about is **constrained models**. For constrained models, _proposed values of position by leapfrog(,etc) can be out of bounds_ (really not sure, but there isn't anything in leapfrog to hold propose values within the constraints), so I think model should also have a `check_constraint` method, which checks whether constraints are satisfied or not. Suggestions ?

khalibartan committed:
Fixes in docstring, renamed theta to position

khalibartan committed:
Arguments names changed

khalibartan commented:
@ankurankan I have changed the arguments names now. Please suggest any other changes, and comment on ambiguous names, etc.

khalibartan commented:
@ankurankan I have changed the arguments names now to make them more descriptive about their purpose . Please suggest changes wherever necessary and point out any ambiguous names, etc. After suggestions are dealt with, I'll move towards optimization.

khalibartan committed:
Added test for errors

khalibartan commented:
@ankurankan I have added all kinds of necessary tests. Please help me out with readability issues.

khalibartan committed:
Optimizing BaseSimulateHamiltonianDynamics

khalibartan commented:
@ankurankan In my perspective PR is ready for final reviews. I have changed names of the arguments according to the context, though there can be issues still please point them out. Also I have optimized things a bit with simulating hamiltonian dynamics, and as expected performance improved.

khalibartan commented:
@ankurankan In my perspective PR is ready for final reviews. I have added all the necessary test and changed names of the arguments according to their context, though there can be issues still, please point them out. Also I have optimized things a bit with simulating hamiltonian dynamics and as expected, we gained a significant performance improvement. (PS: Running the nosetest for it earlier took 14sec and now it takes 8.5 sec on my system)

khalibartan commented on pgmpy/tests/test_inference/test_continuous_sampling.py:
```diff
@@ -0,0 +1,89 @@
+        samples = self.sampler_hmc.sample(theta0, num_adapt=0, num_samples=10000, trajectory_length=3)
+        concatenated_samples = np.concatenate(samples, axis=1)
+        samples_cov = np.cov(concatenated_samples)
+        self.assertTrue(np.linalg.norm(samples_cov - self.test_model.covariance) < 4.0)
```
@ankurankan Here there is an issue, earlier when I pushed my code with tests, `norm of 0.4` was working fine, but things are falling sometimes, and it is getting quite cumbersome because sometimes I increase the norm to make it pass tests and they run fine on my local system but fail while travis runs them. Can suggest something about it ? Should I take a high value of norm ?

ankurankan commented on pgmpy/inference/continuous/base.py:
```diff
@@ -0,0 +1,365 @@
+
+        self.distribution_param = distribution_param
+        self.model = model
+        # The gradient of probability distribution at position
```
@khalibartan Add a blank line above this.

ankurankan commented on pgmpy/inference/continuous/base.py:
```diff
@@ -0,0 +1,365 @@
+        self.model = model
+        # The gradient of probability distribution at position
+        self.grad_log = None
+        # The gradient log of probability distribution at position
```
Blank line above this also

ankurankan commented on pgmpy/inference/continuous/base.py:
```diff
@@ -0,0 +1,365 @@
+
+    def __init__(self, grad_log_pdf, model, position, momentum, stepsize, grad_log_position=None):
+
+        if isinstance(position, (np.matrix, np.ndarray, list, tuple, set, frozenset)):
```
@khalibartan Let's add a separate function to do this check in `utils`.

ankurankan commented on pgmpy/inference/continuous/base.py:
```diff
@@ -0,0 +1,365 @@
+        else:
+            raise TypeError("momentum should be a 1d array type object")
+
+        if not issubclass(grad_log_pdf, BaseGradLogPDF):
```
@khalibartan Use `isinstance`.

khalibartan replied:
@ankurankan I think we cannot use isinstance here, Here is an example:

``` python
>>> class A():
...     def __init__(self):
...         pass
... 
>>> class B(A):
...     def __init__(self):
...         pass
... 
>>> c = B
>>> isinstance(c, A)
False
>>> issubclass(c, A)
True
```

ankurankan commented on pgmpy/inference/continuous/base.py:
```diff
@@ -0,0 +1,365 @@
+        self.model = model
+        self.grad_log_pdf = grad_log_pdf
+        self.grad_log_position = grad_log_position
+        # The new proposed value of position after taking `stepsize` step in time
```
@khalibartan Merge all the 3 comments in one and add a blank line above the comment.

ankurankan commented on pgmpy/inference/continuous/base.py:
```diff
@@ -0,0 +1,365 @@
+    def __init__(self, grad_log_pdf, model, position, momentum, stepsize, grad_log_position=None):
+
+        if isinstance(position, (np.matrix, np.ndarray, list, tuple, set, frozenset)):
+            position = np.array(position).flatten()
```
@khalibartan Why are you flattening it if it's already a 1-D array ?

khalibartan replied:
@ankurankan My bad, I thought I'll be allowing more flexibility of passing any homogenous array type, and will then simply reshape it, in the way I need. I'll change it. This thing will create ambiguity.

ankurankan commented on pgmpy/inference/continuous/base.py:
```diff
@@ -0,0 +1,365 @@
+
+        Returns
+        -------
+        numpy.array: New proposed value of position
```
@khalibartan Add that it returns a tuple. And then explain what are the values.

khalibartan replied:
Okay.

ankurankan commented on pgmpy/inference/continuous/base.py:
```diff
@@ -0,0 +1,365 @@
+        """
+        # Take half step in time for updating momentum
+        momentum_bar = self.momentum + 0.5 * self.stepsize * self.grad_log_position
+        # Take full step in time for updating position position
```
@khalibartan Add a blank line before comment. Looks more cleaner

ankurankan commented on pgmpy/inference/continuous/sampling.py:
```diff
@@ -0,0 +1,564 @@
+
+import numpy as np
+
+from pgmpy.inference.continuous import (LeapFrog, BaseGradLogPDF, BaseSimulateHamiltonianDynamics)
```
@khalibartan No need for the `(`

ankurankan commented on pgmpy/inference/continuous/base.py:
```diff
@@ -0,0 +1,365 @@
+    Inherits pgmpy.inference.base_continuous.BaseGradLogPDF
+    Here model must be an instance of JointGaussianDistribution
+
+    Example
```
@khalibartan Add `Parameters` section.

ankurankan commented on pgmpy/inference/continuous/base.py:
```diff
@@ -0,0 +1,365 @@
+    >>> covariance = np.array([[1, 0.2], [0.2, 7]])
+    >>> model = JointGaussianDistribution(mean, covariance)
+    >>> dist_param = np.array([[0.1], [0.9]])
+    >>> grad_logp, logp = GradLogPDFGaussian(dist_param, model).get_gradient_log_pdf()
```
@khalibartan In the docstring you have mentioned that `dist_param` should be a 1-D array but here you are accepting a 2-D array.

ankurankan commented on pgmpy/inference/continuous/base.py:
```diff
@@ -0,0 +1,365 @@
+    ...         position_bar = self.position + self.stepsize * momentum_bar
+    ...         return position_bar, momentum_bar, grad_log_position
+    >>> import numpy as np
+    >>> pos = np.array([[1], [2]])
```
@khalibartan Why are you accepting 2-D arrays like this everywhere ? I don't think this is intuitive at all. These are all vectors and should be represented by a 1-D array.

ankurankan commented on pgmpy/inference/continuous/sampling.py:
```diff
@@ -0,0 +1,564 @@
+
+        Parameters
+        ----------
+        position0: A 1d array like object
```
@khalibartan Please don't use argument names as position0. `inital_pos` or `start_state` are much more contextual names.

khalibartan replied:
I'll use `start_state`.

khalibartan committed:
Cleaning the HMC code

khalibartan commented:
@ankurankan After exploring things, I came up with the following solution : 
 
- We should take inital_pos as dictionary parameter where dictionary should be something like this:
`{'x' : np.array(), 'y': np.array()}`
Now for this either I can write a class (for now assume its name as DictVector), or we can just take simple dict as input.

- For outputs we can do either of the three things:
We can output a pandas.DataFrame object
We can output a numpy.recarray object
Or If I write DictVector we can output a list of it

What do you suggest ?

khalibartan commented:
Closing. Refer #702 .

khalibartan commented:
Refere #702
